templates/commands/render/templatesource/source.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package templatesource
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+)
+
+var (
+	// This feature is being submitted in pieces, so there's dead code. This
+	// makes the linter stop complaining.
+	_ sourceParser
+	_ templateDownloader
+	_ = realSourceParsers
+	_ = parseSource
+)
+
+// sourceParser is implemented for each particular kind of template source (git,
+// local file, etc.).
+type sourceParser interface {
+	// sourceParse attempts to parse the given src. If the src is recognized as
+	// being downloadable by this sourceParser, then it returns true, along with
+	// a downloader that can download that template.
+	sourceParse(ctx context.Context, src, protocol string) (templateDownloader, bool, error)
+}
+
+// A templateDownloader is returned by a sourceParser, and offers the ability to
+// download a template.
+type templateDownloader interface {
+	// Download downloads this template into the given directory.
+	Download(ctx context.Context, outDir string) error
+}
+
+// realSourceParsers contains the non-test sourceParsers.
+var realSourceParsers = []sourceParser{
+	// This source parser recognizes template sources like
+	// "github.com/myorg/myrepo@v1.2.3" (and variants thereof).
+	&gitSourceParser{
+		re: regexp.MustCompile(
+			`^` + // Anchor the start, must match the entire input
+				`(?P<host>github\.com|gitlab\.com)` + // The domain names of known git hosting services
+				`/` +
+				`(?P<org>[a-zA-Z0-9_-]+)` + // the github org name, e.g. "abcxyz"
+				`/` +
+				`(?P<repo>[a-zA-Z0-9_-]+)` + // the github repo name, e.g. "abc"
```

Comment:
It's not very clear to me why these sub matching are needed. Not saying we need to but curious about what stops us from supporting something like `https://${other_git_server}@latest`

Reply:
I'll send you the design doc in chat 😀

Reply:
Actually, I can just answer here. One of the things we're doing in this series of PRs is creating a namespace for versioned templates that is separate from the transport details (https or ssh). This lets us do things like "template X depends on template Y", and we can refer to template Y without being overly specific about the details of the transport protocol that we use to fetch template Y. It's similar to `go get` in that way.

Reply:
Understood. But what I was asking about was the host/org/repo part. Why do we need to know about them? From your description, it doesn't sound like we need them for the template dependency?

Reply:
It's just to verify that the structure is what we expect. If somebody gives a template source of `github.com/foo` (org name with no repo name), this lets us detect that it's not a valid remote git repo, because it doesn't match the regex. We can continue with other sourceParsers and try treating it as a local file or $FUTURE_TEMPLATE_SOURCE_TYPE.

We could create one big capturing group for "host plus org plus repo but not including subdirectory or version", but that might be harder to understand because that's a confusing union of concepts for the reader to grok.

Reply:
I see. I mainly question whether we really need to care about the format. Would it still work if it's not github/gitlab format?

Reply:
If it's not github/gitlab, then it won't work, but that's by design. In that case, we'll continue with the next sourceParser and attempt to treat the input as a local file rather than a git repo.